core_types/src/eav.rs

```diff
@@ -122,88 +176,117 @@ pub trait EntityAttributeValueStorage: objekt::Clone + Send + Sync + Debug {
         entity: Option<Entity>,
         attribute: Option<Attribute>,
         value: Option<Value>,
-    ) -> Result<HashSet<EntityAttributeValue>, HolochainError>;
+        index_query: IndexQuery,
+    ) -> Result<BTreeSet<EntityAttributeValueIndex>, HolochainError>;
 }
 
 clone_trait_object!(EntityAttributeValueStorage);
 
 #[derive(Clone, Debug)]
-pub struct ExampleEntityAttributeValueStorageNonSync {
-    storage: HashSet<EntityAttributeValue>,
+pub struct ExampleEntityAttributeValueStorage {
+    storage: Arc<RwLock<BTreeSet<EntityAttributeValueIndex>>>,
 }
-
-impl ExampleEntityAttributeValueStorageNonSync {
-    pub fn new() -> ExampleEntityAttributeValueStorageNonSync {
-        ExampleEntityAttributeValueStorageNonSync {
-            storage: HashSet::new(),
+impl ExampleEntityAttributeValueStorage {
+    pub fn new() -> ExampleEntityAttributeValueStorage {
+        ExampleEntityAttributeValueStorage {
+            storage: Arc::new(RwLock::new(BTreeSet::new())),
         }
     }
+}
+
+pub fn increment_key_till_no_collision(
+    mut eav: EntityAttributeValueIndex,
+    map: BTreeSet<EntityAttributeValueIndex>,
+) -> HcResult<EntityAttributeValueIndex> {
+    if map
+        .iter()
+        .filter(|e| e.index == eav.index())
+        .collect::<BTreeSet<&EntityAttributeValueIndex>>()
+        .len()
+        > 0
+    {
+        let timestamp = eav.clone().index + 1;
+        eav.set_index(timestamp);
+        increment_key_till_no_collision(eav, map)
```

Comment:
how does this preserve ordering better than generating a random squuid in a threaded situation?

Reply:
Guards on eav so only one write per time, timestamps are sorted by i64 thus ordered

Reply:
so then i have the inverse question, why do we need to call it recursively

if this is one write at a time we should just be able to to `+ 1` against the current highest value if it equals the proposed index

Reply:
This is a more fireproof way of incase you have multiple collisions on the nanoseconds.
value 1 = 0001, value 2 =0001, value 3 = 0001

If both of these are created at the same time, value 2 would be 0002 if I add only one. If it happens to value 3 and we add only 1 it would happen override value 2 because the timestamp would be 0002. That is why we need to iterate until we find no collisions :)